File: knowledgeplan-frontend/src/types/goal.ts
// import { Project } from './project'; // Assuming relationship

// Basic Goal/OKR structure based on common fields and potential relationships
// Mirroring backend schemas/goal.py if possible

// import uuid from 'uuid'; // Remove unused import

// Original structure - Keeping simple Goal interface for now
export interface Goal {
    id: string; // Assuming UUID
    name: string; // Use title preferably from GoalRead
    description?: string | null;
    status?: string | null; // Or an enum (e.g., On Track, At Risk)
    progress?: number | null; // e.g., 0-100
    parent_id?: string | null; // For parent/child OKR relationships
    project_id?: string | null; // Link to a specific project
}

// Added based on backend schemas/goal.py GoalRead
// Use this for the main entity fetch
export interface GoalRead {
    id: string;
    tenant_id: string;
    title: string;
    description?: string | null;
    type?: string | null; // Enum: Enterprise/Dept/Team
    parent_id?: string | null; 
    status?: string | null;
    progress?: number | null;
    dueDate?: string | null; // Date as string
    created_at: string;
    updated_at: string;
    properties?: Record<string, unknown> | null;
}

// Explicit definition matching backend GoalReadMinimal schema
export enum GoalTypeEnum {
    ENTERPRISE = "enterprise",
    DEPARTMENT = "department",
    TEAM = "team",
    INDIVIDUAL = "individual"
}

export interface GoalReadMinimal {
    id: string; // UUID
    tenant_id: string;
    title: string;
    description?: string | null;
    type: GoalTypeEnum; // Use the Enum here
    parent_id?: string | null;
    status?: string | null;
    progress?: number | null;
    // Use due_date to match backend query result if different from GoalRead
    due_date?: string | null; 
    created_at: string;
    updated_at: string;
}

// Keep Create/Update simple for now if not used heavily
export interface GoalCreate {
    title: string;
    description?: string | null;
    type?: string | null;
    parent_id?: string | null;
    // ... other fields needed on creation
}

export interface GoalUpdate {
    title?: string | null;
    description?: string | null;
    type?: string | null;
    status?: string | null;
    progress?: number | null;
    dueDate?: string | null;
    parent_id?: string | null;
    // ... other fields allowed for update
} 

File: knowledgeplan-backend/app/schemas/goal.py
import uuid
from datetime import datetime, date
from typing import Any, Optional
from enum import Enum as PyEnum
from uuid import UUID

from pydantic import BaseModel, Field, UUID4

class GoalTypeEnum(str, PyEnum):
    ENTERPRISE = "enterprise"
    DEPARTMENT = "department"
    TEAM = "team"
    INDIVIDUAL = "individual" # Maybe add later

# Shared properties
class GoalBase(BaseModel):
    title: str
    description: Optional[str] = None
    type: Optional[str] = None # Consider Enum later: Enterprise/Dept/Team
    status: Optional[str] = None
    progress: Optional[int] = Field(None, ge=0, le=100)
    dueDate: Optional[date] = Field(None, alias='dueDate') # Match frontend if needed
    parent_id: Optional[UUID] = None
    # properties: Optional[dict] = {} # Replaced with JSON type below

    class Config:
        allow_population_by_field_name = True # Allow using dueDate alias

# Properties to receive via API on creation
class GoalCreate(GoalBase):
    pass # Inherits all from GoalBase for now

# Properties to receive via API on update
class GoalUpdate(GoalBase):
    # Make all fields optional for update
    title: Optional[str] = None
    description: Optional[str] = None
    type: Optional[str] = None
    status: Optional[str] = None
    progress: Optional[int] = Field(None, ge=0, le=100)
    dueDate: Optional[date] = Field(None, alias='dueDate')
    parent_id: Optional[UUID] = None
    # properties: Optional[dict] = None

# Properties shared by models stored in DB
class GoalInDBBase(GoalBase):
    id: UUID
    tenant_id: UUID
    created_at: datetime
    updated_at: datetime
    # properties: Optional[Any] = None # Use JSON type from SQLAlchemy

    # Use Pydantic V2 config
    model_config = {
        "from_attributes": True,
        "populate_by_name": True, # Replaces allow_population_by_field_name
    }

# Properties to return to client
class GoalRead(GoalInDBBase):
    pass # Inherits all needed fields

# Minimal Goal details (e.g., for lists where properties are excluded)
class GoalReadMinimal(BaseModel):
    # Match the columns selected in the specific query
    id: UUID
    tenant_id: UUID
    title: str
    description: Optional[str] = None
    type: GoalTypeEnum
    parent_id: Optional[UUID] = None
    status: Optional[str] = None
    progress: Optional[int] = None
    due_date: Optional[date] = None
    # Exclude properties
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

# Properties stored in DB
class GoalInDB(GoalInDBBase):
    pass 